frame/mosaic-vault/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+	
+	#[pallet::config]
+    pub trait Config: frame_system::Config {
+
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Currency: Transfer<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>
+		     + Mutate<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>
+			 + Inspect<Self::AccountId, Balance = Self::Balance, AssetId = Self::AssetId>;
+
+		type Convert: Convert<Self::Balance, u128> + Convert<u128, Self::Balance>;
+
+		type Balance: Parameter 
+		    + Member 
+			+ AtLeast32BitUnsigned 
+			+ Codec 
+			+ Default 
+			+ Copy 
+			+ MaybeSerializeDeserialize 
+			+ Debug 
+			+ MaxEncodedLen 
+			+ TypeInfo 
+			+ CheckedSub 
+			+ CheckedAdd 
+			+ Zero 
+			+ PartialOrd;
+
+		type Nonce:  Parameter + Member + AtLeast32BitUnsigned + Codec + Default + Copy + MaybeSerializeDeserialize + Debug + MaxEncodedLen + TypeInfo + CheckedSub + CheckedAdd;//+ From<u8>;
+
+		type TransferDelay:  Parameter + Member + AtLeast32BitUnsigned + Codec + Default + Copy + MaybeSerializeDeserialize + Debug + MaxEncodedLen + TypeInfo;
+
+		type VaultId: Clone 
+		    + Codec 
+			+ Debug 
+			+ PartialEq 
+			+ Default 
+			+ Parameter;
+
+		type Vault: StrategicVault<
+			VaultId = Self::VaultId,
+			AssetId = <Self as Config>::AssetId,
+			Balance = Self::Balance,
+			AccountId = Self::AccountId,>;
+
+		type AssetId: FullCodec
+		     + Eq
+			 + PartialEq
+			 + Copy
+			 + MaybeSerializeDeserialize
+			 + Debug
+			 + Default
+			 + TypeInfo;
+		
+		type RemoteAssetId: FullCodec
+			 + Eq
+			 + PartialEq
+			 + Copy
+			 + MaybeSerializeDeserialize
+			 + Debug
+			 + Default
+			 + TypeInfo;
+		
+		type RemoteNetworkId: FullCodec
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ TypeInfo;
+
+		type DepositId: FullCodec
+			+ Eq
+			+ PartialEq
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Debug
+			+ Default
+			+ TypeInfo;
+
+		type RelayerOrigin: EnsureOrigin<Self::Origin>;
+
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+
+		#[pallet::constant]
+		type FeeFactor: Get<Self::Balance>;
+
+		#[pallet::constant]
+		type ThresholdFactor: Get<Self::Balance>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		type FeeAddress: Get<Self::AccountId>;
+
+		type BlockTimestamp: UnixTime;
+
+		type MaxFeeDefault: Get<Self::Balance>;
+
+		type MinFeeDefault: Get<Self::Balance>;
+	}
+	#[derive(Encode, Decode, Default, Debug, PartialEq, TypeInfo)]
+	pub struct DepositInfo<AssetId, Balance > {
+        pub asset_id: AssetId,
+		pub amount: Balance,
+	}  
+	
+	#[pallet::storage]
+	#[pallet::getter(fn remote_asset_id)]
+    pub(super) type RemoteAssetId<T: Config> = StorageDoubleMap<_, Blake2_128Concat, T::RemoteNetworkId, Blake2_128Concat, T::AssetId, T::RemoteAssetId, ValueQuery>;
```

Comment:
may be use OptionQuery? so than option can be get and no need to double call storage to check if asset exists